packages/xstate-graph/src/graph.ts

```diff
@@ -171,182 +167,141 @@ export function getAdjacencyMap<
     }
   }
 
-  findAdjacencies(node.initialState);
+  findAdjacencies(machine.initialState as TState);
 
   return adjacency;
 }
 
-export function getShortestPaths<
-  TContext = DefaultContext,
-  TEvent extends EventObject = EventObject
->(
-  machine: StateMachine<TContext, any, TEvent, any, any, any, any>,
-  options?: ValueAdjMapOptions<TContext, TEvent>
-): StatePathsMap<TContext, TEvent> {
-  if (!machine.states) {
-    return EMPTY_MAP;
+const defaultMachineStateOptions: TraversalOptions<State<any, any>, any> = {
+  serializeState: serializeMachineState,
+  serializeEvent,
+  getEvents: (state) => {
+    return state.nextEvents.map((type) => ({ type }));
   }
-  const optionsWithDefaults = getValueAdjMapOptions(options);
+};
 
-  const adjacency = getAdjacencyMap<TContext, TEvent>(
-    machine,
-    optionsWithDefaults
+export function getShortestPlans<TMachine extends AnyStateMachine>(
+  machine: TMachine,
+  options?: Partial<TraversalOptions<AnyState, EventObject>>
+): Array<StatePlan<AnyState, EventObject>> {
+  const resolvedOptions = resolveTraversalOptions(
+    options,
+    defaultMachineStateOptions
+  );
+  return traverseShortestPlans(
+    {
+      transition: (state, event) => machine.transition(state, event),
+      initialState: machine.initialState
+    },
+    resolvedOptions
   );
+}
+
+export function traverseShortestPlans<TState, TEvent extends EventObject>(
+  behavior: SimpleBehavior<TState, TEvent>,
+  options?: Partial<TraversalOptions<TState, TEvent>>
+): Array<StatePlan<TState, TEvent>> {
+  const optionsWithDefaults = resolveTraversalOptions(options);
+  const { serializeState } = optionsWithDefaults;
+
+  const adjacency = performDepthFirstTraversal(behavior, optionsWithDefaults);
 
   // weight, state, event
   const weightMap = new Map<
-    string,
-    [number, string | undefined, string | undefined]
+    SerializedState,
+    [number, SerializedState | undefined, TEvent | undefined]
   >();
-  const stateMap = new Map<string, State<TContext, TEvent>>();
-  const initialVertex = optionsWithDefaults.stateSerializer(
-    machine.initialState
-  );
-  stateMap.set(initialVertex, machine.initialState);
+  const stateMap = new Map<SerializedState, TState>();
+  const initialSerializedState = serializeState(behavior.initialState, null);
+  stateMap.set(initialSerializedState, behavior.initialState);
 
-  weightMap.set(initialVertex, [0, undefined, undefined]);
-  const unvisited = new Set<string>();
+  weightMap.set(initialSerializedState, [0, undefined, undefined]);
+  const unvisited = new Set<SerializedState>();
   const visited = new Set<string>();
```

Comment:
If the `unvisited` type has been changed here then the same thing should happen for `visited` since both are highly related (if i read this right):
```suggestion
  const visited = new Set<SerializedState>();
```